### PR TITLE
[bugfix] make Mnist training converge to baseline 

### LINKED ISSUE #212: invalid output number of `torch.eq(t1, t2).sum()`
From the MNIST example of TorchDisc, the program outputs an invalid test acc number:

``` text
Test set: Average loss: 0.3399, Accuracy: 281447840135434/10000 (2814478401354%)
``` 

A tiny program to reproduce this bug:

```python
import torch
import _torch_disc as disc
disc._ltc_init_disc_backend()

def func():
    t1 = torch.tensor([[1, 2], [3, 4]])
    t2 = torch.tensor([[1, 2], [5, 6]])
    t1, t2 = t1.to('lazy'), t2.to('lazy')
    t3 = torch.eq(t1, t2)
    return t3.sum()

print(func())
```

outputs:

``` text
tensor(3328288142057042689, device='lazy:0')
```

### LINKED ISSUE #215: `addmm` op caused an unchanged loss
In #207, I added `addmm` op to the blacklist to lowing to Disc, because this caused the training to remain unchanged. It seems a bug, we should debug and fix it.

### PR DESCRIPTION
**Pull request by Yancey1989**
fixed #215 
fixed #212 

This PR makes Mnist converge to baseline with TorchDisc, and also adds a unit test to make sure the baseline model (Mnist)
 always correct in the future.

image

### CODE REVIEW

File: torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp [hunk 1]
```diff
@@ -24,36 +24,38 @@ using namespace ::torch::jit;
...
 std::vector<Node*> FakeCluster(const std::shared_ptr<Graph>& graph) {
   std::vector<Node*> nodes;
-  for (auto node : graph->nodes()) {
-    if (torch::blade::IsMlirMhloSupported(*node) &&
-        node->kind() != prim::Constant) {
-      if (node->kind() == aten::addmm) continue;
-      nodes.push_back(node);
-    }
-  }
+  std::copy_if(graph->nodes().begin(), graph->nodes().end(),
```
Comment (bddppq) [inline torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp:27]: `cbegin`

File: torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp [hunk 2]
```diff
@@ -24,36 +24,38 @@ using namespace ::torch::jit;
...
+//      %p2.1: Tensor):
+//    %1 int = aten::item(%p2.1)
+//    %2 Tensor = aten::add(%p0.1, %p1.1, %1)
+//    return %2
 void CastBoundaryScalarToTensor(Graph* disc_graph, size_t i) {
   auto new_input = disc_graph->insertInput(
       i, c10::string(disc_graph->inputs()[i]->debugName() + ".1"));
+  // TODO(disc): derivate input scalar type
+  new_input->setType(
+      TensorType::create(at::ScalarType::Int, c10::nullopt, 0, false));
```
Comment (bddppq) [inline torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp:58]: Is this pass currently only used in the experimental training support? (otherwise blindly setting dtype to int would affect correctness of other cases)

File: docker/dev/Dockerfile.torch-disc [hunk 3]
```diff
@@ -8,5 +8,5 @@ RUN apt-get update -y && \
     rm /usr/bin/python && \
     ln -s /usr/bin/python3.8 /usr/bin/python && \
     python -m pip install --upgrade pip && \
-    python -m pip install cpython pyyaml typing_extensions virtualenv numpy
-RUN python -m pip install https://bladedisc-ci.oss-cn-hongkong.aliyuncs.com/download/torch-ltc/torch-1.12.0a0%2Bgitd9896b8-cp38-cp38-linux_x86_64.whl
\ No newline at end of file
+    python -m pip install cpython pyyaml typing_extensions virtualenv numpy torchvision
+RUN python -m pip uninstall pytorch && python -m pip install https://bladedisc-ci.oss-cn-hongkong.aliyuncs.com/download/torch-ltc/torch-1.12.0a0%2Bgitd9896b8-cp38-cp38-linux_x86_64.whl
```
[the remark this answers is not in the document]
Comment (qiuxiafei) [reply docker/dev/Dockerfile.torch-disc:12]: what magic is in this torch wheel?

**Review by qiuxiafei — COMMENTED** [review]
Another thing off the topic of this CR.
Maybe is not a good idea to introduce the concept of `torch_disc` to developers, we've already disc/blade/torch-blade ... My immature thought is to unify `torch_disc` alone with `mhlo_builder` and some code under `pytorch_blade/src/compiler/mlir` to be `torch_bridge` of DISC. Any comments ? @fortianyou

File: docker/dev/Dockerfile.torch-disc (hunk 3 above)
[the remark this answers is not in the document]
Comment (Yancey1989) [reply docker/dev/Dockerfile.torch-disc:12]: > we probably should build torchvision with pinned git sha as well

Agree with that, maybe we can do that in the future PR because we just use the Mnist dataset preprocessing, and I will leave an issue to trace it.

[the remark this answers is not in the document]
Comment (Yancey1989) [reply docker/dev/Dockerfile.torch-disc:12]: > what magic is in this torch wheel?

build from the [lazy_tensor_staging branch of PyTorch](https://github.com/pytorch/pytorch/tree/lazy_tensor_staging).

File: torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp (hunk 2 above)
Comment (Yancey1989) [reply torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp:58]: thanks for your reminder, It's not only the experimentation, and hard code to int would cause unexpected behavior and be hard to debug, I will update this PR later. :)

**Comment by fortianyou** [comment]
> torch_bridge

I don't think we need a `torch_bridge`, since `torch_blade` satisfies what we need. But I approve we should introduce new concepts carefully.

File: torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp (hunk 2 above)
Comment (Yancey1989) [reply torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp]: done, add `DerivateScalarType` function.

File: torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp (hunk 1 above)
Comment (Yancey1989) [reply torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp:28]: may not, we need the mutable node pointer, because it would be destroyed later.

https://github.com/alibaba/BladeDISC/blob/41dfc36902cf63c0e488c970da7b8a4e7763ce9a/torch_disc/torch_disc/csrc/disc_compiler/passes/register_disc_class.cpp#L99

**Comment by qiuxiafei** [comment]
Well, `torch_bridge` may not be accurate. `torch_blade` is sufficient enough as the top-level entry from pytorch to DISC. As for `torch_bridge`, it should be part of DISC and under DISC's directory in the future, a top-level `mhlo_builder ` looks weird.

File: torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp (hunk 2 above)
Comment (Yancey1989) [reply torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp]: done, added stricter inspection on input dtype and scalar type derivation using `c10::scalarTypeFromJitType`.

**Review by fortianyou — APPROVED** [review]
LGTM.

File: torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp [hunk 4]
```diff
@@ -77,7 +90,8 @@ void CastGraphInputsToTensor(const std::shared_ptr<Graph>& graph,
       sub_graph->replaceInput(i, cast_tensor->output());
 
       // TODO(Yancey1989): cast output
```
Comment (bddppq) [inline torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp:92]: TODO no longer valid?

File: torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp [hunk 5]
```diff
@@ -24,36 +24,49 @@ using namespace ::torch::jit;
...
+//      %p1.1: Tensor,
+//      %p2.1: Tensor):
+//    %1 int = aten::item(%p2.1)
+//    %2 Tensor = aten::add(%p0.1, %p1.1, %1)
+//    return %2
+void CastBoundaryScalarToTensor(Graph* disc_graph, size_t i,
+                                at::ScalarType& typ) {
   auto new_input = disc_graph->insertInput(
       i, c10::string(disc_graph->inputs()[i]->debugName() + ".1"));
+  new_input->setType(TensorType::create(typ, c10::nullopt, 0, false));
```
Comment (bddppq) [inline torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp:69]: Should `item_node->output()` be set to the same type? (aka line 73)

**Comment by Yancey1989** [comment]
@fortianyou @qiuxiafei thanks for your comments on director structure, I agree that we should re-design the top-level dir to make this repo more concise. Some rough ideas on the top-level directory:

- the repo name is `BladeDISC`, so maybe we can remove some redundancy `blade` keyword, using `tensorflow`, `pytorch` instead of `tensorflow_blade`, `pytorch_blade`...
- a lot of repo use programing languages to organize dirs, e.g. [pytorch](https://github.com/pytorch/pytorch/tree/master/torch) , [tensorflow](https://github.com/tensorflow/tensorflow/tree/master/tensorflow), or even [sqlflow](https://github.com/sql-machine-learning/sqlflow)

So the director structure can be re-orgnized as the following:

``` text
docker        // Docker program
scripts         // Some Python/Bash scripts
disc     
    |- mlir   // compiler core
    |- cc
    |- ral   // RAL
tensorflow
    |- bridge_op    // tao_launch operator kernel
    |- passes       // some passes
pytorch
    |- mhlo         // mhlo conversion
    |- ltc_bridge   // lazy tensor core bridge on training
    |- script_module // TorchScript module bridge on inference
```

BTW,  the discussion about director structure is not related on this pull request, maybe we can move to an issue

**Comment by qiuxiafei** [comment]
Yes, let go to some issue ~

File: torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp (hunk 5 above)
Comment (Yancey1989) [reply torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp:80]: Done. Thanks, we should do that.

File: torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp (hunk 4 above)
Comment (Yancey1989) [reply torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp:102]: I think it's valid in the future, the output value in a disc module may be Scalar type, we should cast them to Tensor inner the module and cast them back to Scalar outer the module, but we have not met this scene, will fix this TODO in the future. And #176 introduced some background about the casting operation.

Comment (bddppq) [reply torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp:102]: I see

File: torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp [hunk 6]
```diff
@@ -24,40 +24,63 @@ using namespace ::torch::jit;
...
-  }
+    return false;
+  };
+
+  std::copy_if(graph->nodes().begin(), graph->nodes().end(),
+               std::back_inserter(nodes), is_disc_compilable);
   return nodes;
 }
 
+c10::TypePtr getScalarTypePtr(at::ScalarType& typ) {
```
Comment (bddppq) [inline torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp:47]: Please check if there is such a mapping somewhere in our project already. cc @fortianyou

**Review by bddppq — APPROVED** [review]
LGTM
